api/tests/e2e/helpers/fail_handler.go

```diff
@@ -0,0 +1,107 @@
+package helpers
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+)
+
+const (
+	logTailLines = 50
+)
+
+type podContainerDescriptor struct {
+	Namespace  string
+	LabelKey   string
+	LabelValue string
+	Container  string
+}
+
+func E2EFailHandler(message string, callerSkip ...int) {
+	config, err := controllerruntime.GetConfig()
+	Expect(err).NotTo(HaveOccurred())
+
+	clientset, err := kubernetes.NewForConfig(config)
+	Expect(err).NotTo(HaveOccurred())
+
+	printPodsLogs(clientset, []podContainerDescriptor{
+		{
+			Namespace:  "cf-k8s-api-system",
+			LabelKey:   "app",
+			LabelValue: "cf-k8s-api",
+			Container:  "cf-k8s-api",
+		},
+		{
+			Namespace:  "cf-k8s-controllers-system",
+			LabelKey:   "control-plane",
+			LabelValue: "controller-manager",
+			Container:  "manager",
+		},
+	})
+
+	Fail(message, callerSkip...)
+}
+
+func printPodsLogs(clientset kubernetes.Interface, podContainerDescriptors []podContainerDescriptor) {
+	for _, desc := range podContainerDescriptors {
+		pods, err := getPods(clientset, desc.Namespace, desc.LabelKey, desc.LabelValue)
+		if err != nil {
+			fmt.Fprintf(GinkgoWriter, "Failed to get pods with label %s=%s: %v\n", desc.LabelKey, desc.LabelValue, err)
+		}
+
+		for _, pod := range pods {
+			log, err := getSinglePodLog(clientset, pod, desc.Container)
+			if err != nil {
+				fmt.Fprintf(GinkgoWriter, "Failed to get logs for pod %s: %v\n", pod.Name, err)
+
+				continue
+			}
+
+			fmt.Fprintf(GinkgoWriter,
+				"\n\n===== Logs for pod %q (last %d lines) =====\n%s\n==============================================\n\n",
+				pod.Name,
+				logTailLines,
+				log)
+		}
+	}
+}
+
+func getPods(clientset kubernetes.Interface, namespace, labelKey, labelValue string) ([]corev1.Pod, error) {
+	pods, err := clientset.CoreV1().Pods(namespace).List(context.Background(), metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("%s=%s", labelKey, labelValue),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return pods.Items, nil
+}
+
+func getSinglePodLog(clientset kubernetes.Interface, pod corev1.Pod, container string) (string, error) {
+	podLogOpts := corev1.PodLogOptions{TailLines: int64ptr(logTailLines), Container: container}
+	req := clientset.CoreV1().Pods(pod.Namespace).GetLogs(pod.Name, &podLogOpts)
+
+	logStream, err := req.Stream(context.Background())
+	if err != nil {
+		return "", err
+	}
+	defer logStream.Close()
+
+	logBytes, err := ioutil.ReadAll(logStream)
```

Comment:
Unrelated to this PR, but wanted to run a thought by y'all. 

Do you think this approach could be used to implement the log-cache API READ envelope endpoint? It seems promising... pod logs are a subresource of `Pods` so could be RBAC-ed. That'd handle the access control question. And we could then stream them into something that converts them into the loggregator envelope format

Right now we've just got a hard-coded implementation that always returns an empty array of envelopes, but we're thinking about what the future of this looks like.
https://github.com/cloudfoundry/cf-k8s-controllers/issues/222

Reply:
I doubt how reliable would such an implementation be. 
Pods are being managed by K8s, meaning that when a pod is killed, its logs are lost. Therefore, if an app is constantly crashing, its logs are going to be immediately lost after the crash so and endpoint that just reads running pods logs would not be much helpful.

Reply:
Right now it's `cf logs` returns no logs so we're looking for ways for it to return _some_. I think in the scenario you describe even cf-for-k8s with it's log-caches and fluentds had trouble with it since they didn't get logs before the containers crashed.

Rather than re-implement the whole CF logging subsystem, though, I anticipate advising operators to use existing K8s log aggregators for longer-term storage.

Reply:
I don't understand how log streaming works in log-cache given their response is a JSON object 🤔